### PR TITLE
(WIP) Improve BMS faults test coverage

### PR DESCRIPTION
**Pull request by gtaharaedmonds**
### Summary
Improve tests related to BMS faults.

NOTE: BMS doesn't yet open contactors before it faults. (TODO, needs tests too)

### Changelist 
- If AIR- opens during precharge, BMS now enters fault state
- Centralize all BMS missing heartbeat checks to all states file
- BMS fault tests are generally more rigorous: Instead of just checking we enter fault state when fault occurs, we now test that we stay in fault state indefinitely until fault clears. Also, we now test that we'll transition back to init state after fault clears. 

Added 3x new tests: 
- Test state machine behavior for precharge faults
- Test missing heartbeat will cause us to enter fault state
- Test entering fault state will make a call to disable BMS OK

### Testing Done
- This PR is mostly test changes so only unit testing

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: boards/shared/Src/App/App_SharedHeartbeatMonitor.c [hunk 1]
```diff
@@ -42,7 +42,7 @@ bool App_SharedHeartbeatMonitor_Tick(struct HeartbeatMonitor *const heartbeat_mo
 
     if ((current_ms - heartbeat_monitor->previous_timeout_ms) >= heartbeat_monitor->timeout_period_ms)
     {
-        heartbeat_monitor->previous_timeout_ms += heartbeat_monitor->timeout_period_ms;
+        heartbeat_monitor->previous_timeout_ms = current_ms;
```
Comment (gtaharaedmonds) [inline boards/shared/Src/App/App_SharedHeartbeatMonitor.c:46]: The previous line was making it hard to run tests, and I think the new line makes more sense anyway? Could you a double check here

File: boards/BMS/Test/Src/Test_Faults.cpp [hunk 2]
```diff
@@ -475,4 +579,120 @@ TEST_F(BmsFaultTest, check_precharge_fault)
...
+
+    // Set cell voltage critically high and confirm fault is set
+    set_cell_voltage((AccumulatorSegment)0U, 0U, MAX_CELL_VOLTAGE + 0.1f);
+    LetTimePass(state_machine, 20);
+    ASSERT_EQ(App_GetFaultState(), App_SharedStateMachine_GetCurrentState(state_machine));
+
+    // BMS OK status should be disabled upon entering fault state, which is latched in hardware
+    ASSERT_EQ(enable_bms_ok_fake.call_count, 0);
+    ASSERT_EQ(disable_bms_ok_fake.call_count, 1);
+}
```
Comment (gtaharaedmonds) [inline boards/BMS/Test/Src/Test_Faults.cpp:696]: Not super rigorous but confirms we at least attempt to latch BMS not OK status upon entering fault state

File: boards/BMS/Src/App/states/App_FaultState.c [hunk 3]
```diff
@@ -32,12 +33,16 @@ static void FaultStateRunOnTick100Hz(struct StateMachine *const state_machine)
...
-    App_CanAlerts_SetFault(BMS_FAULT_MISSING_HEARTBEAT, hb_ok);
+    const bool dcm_ok        = !App_CanAlerts_BoardHasFault(DCM_ALERT_BOARD);
+    const bool fsm_ok        = !App_CanAlerts_BoardHasFault(FSM_ALERT_BOARD);
+    const bool pdm_ok        = !App_CanAlerts_BoardHasFault(PDM_ALERT_BOARD);
+    const bool dim_ok        = !App_CanAlerts_BoardHasFault(DIM_ALERT_BOARD);
+    const bool no_can_alerts = dcm_ok && fsm_ok && pdm_ok && dim_ok;
 
-    if (acc_fault_cleared && ts_fault_cleared && is_air_negative_open && hb_ok)
+    if (acc_fault_cleared && ts_fault_cleared && is_air_negative_open && hb_ok && precharge_ok && no_can_alerts)
     {
```
Comment (gtaharaedmonds) [inline boards/BMS/Src/App/states/App_FaultState.c:46]: can't leave fault state until other boards clear their faults

File: boards/shared/Src/App/App_SharedHeartbeatMonitor.c (hunk 1 above)
Comment (will-chaba) [reply boards/shared/Src/App/App_SharedHeartbeatMonitor.c:46]: can't see why the functionality would be affected with this change, so if it helps with testing looks fine to me

**Review by will-chaba — APPROVED** [review]
LGTM, think I reviewed some of these changes as part of the simple torque requests pr as well

**Pull request merged**